### PR TITLE
move remaining state to db, allow multiple build servers

### PR DESCRIPTION
**Pull request by syphar**
This is a possible approach for #795, without #1011, but including the possibility to run multiple build-servers.  

It contains: 
- moving the queue-lock to the database 
- moving the _last seen reference_ into the database, so we can recreate the registry watcher from scratch without any issues. 
- an argument to run a registry watcher via `start-registry-watcher`. With multiple build-servers we still should only run one registry watcher 
- the repository-stats-updater can be run with a scheduled ECS job (via `database update-repository-fields`), or optionally in the registry watcher process (via `--repository-stats-updater=enabled`). 
- a separate build-server (via `start-build-server`) which can be run multiple times. A queued crate will only be picked up by one of the servers. 

we already can run a separate webserver as often as we want with `start-web-server`, of course using the proper connection pooling limits. 

### testing 
Since test coverage for the build-process and queue is limited I might miss edge-cases. I'm happy to fix any issues that might come up.

### CODE REVIEW

**Comment by syphar** [comment]
@jyn514 @Nemo157 it would be awesome to get feedback on this

**Comment by Nemo157** [comment]
At first skim it all looks great to me. I'll take a deeper look again later, and I'll have to read the postgres docs around the locking.

File: src/build_queue.rs
Comment (jyn514) [inline src/build_queue.rs:58]: Shouldn't this return an error instead of discarding it?

File: src/build_queue.rs
Comment (jyn514) [inline src/build_queue.rs:317]: It worries me that this isn't atomic. I worry that we'll:
- update this in the local git repo
- fail to update it in the database
- end up building the same list of crates twice

I guess that's not the end of the world. But it would be nice to add some more error logging, and to update it in the database before updating the git repo.

File: src/utils/daemon.rs
Comment (jyn514) [inline src/utils/daemon.rs:35]: Can you add a comment that there should only be one registry watcher running at a time? It confused me for a bit why there wasn't any locking.

In practice I think it should be ok - the worst that happens is we'll try and insert the same crates twice, and the second insert will be ignored by the database.

File: src/utils/queue_builder.rs
Comment (jyn514) [inline src/utils/queue_builder.rs:22]: No longer a file.
```suggestion
            warn!("Build queue is locked, skipping building new crates");
```

File: src/utils/queue_builder.rs
Comment (jyn514) [inline src/utils/queue_builder.rs:41]: There's a comment below that says "If we panic here something is really truly wrong and trying to handle the error won't help". That no longer seems true to me, since the database could be down or whatever. But I don't know how to handle the error. Maybe we should treat it the same as the rest of the errors, exit the build loop and report it?

**Review by jyn514 — APPROVED** [review]
This is awesome ❤️ thank you!!

**Comment by syphar** [comment]
> At first skim it all looks great to me. I'll take a deeper look again later, and I'll have to read the postgres docs around the locking.

On top of the docs [this article](https://www.enterprisedb.com/blog/what-skip-locked-postgresql-95) is also helpful. 

I ran some local tests (replacing the build with a `sleep`) and 3 build-servers correctly picked up queued crates

File: src/build_queue.rs
Comment (syphar) [reply src/build_queue.rs:58]: I was kind of coming from the `watch_registry` method where I want to keep the registry watcher running, but locked, after this happened. 

So we would then manually remove / fix the ref and unlock the queue. 

But I can see that this could be confusing behavior for `last_seen_reference` and moved the logic into `watch_registry`

File: src/utils/daemon.rs
Comment (syphar) [reply src/utils/daemon.rs:45]: I added a note

File: src/utils/queue_builder.rs
Comment (syphar) [reply src/utils/queue_builder.rs:41]: For a moment I thought you meant exiting = `break`, but I think you mean `continue` like with the other errors. 

I don't know which cases can lead to panics in the build. I remember we still had some places sprinkled with `.unwrap` or `.expect` around database methods, where trying again could be valid. 

Checking this I also saw that the `?` when checking the queue lock should be removed to match the error-handling in other cases in that loop.

Comment (syphar) [reply src/utils/queue_builder.rs:49]: I updated the logic , only reporting and `continue` also for panics

File: src/build_queue.rs
Comment (syphar) [reply src/build_queue.rs:306]: I changed the order of these, first setting the ref in the db, then in the repo. 

What additional logging to you imagine? 

At the call-site of `get_new_crates` ( now in `watch_registry`) all errors will be logged already.

File: src/utils/queue_builder.rs
Comment (Nemo157) [reply src/utils/queue_builder.rs:49]: IIUC, the failure case here if it's the crate causing the failure will just be that we retry building the top crate from the queue continuously every 60 seconds (because the panic bypasses incrementing the attempt counter); until someone manually removes that crate from the queue (which would have been the same resolution before). If it's something transient like the database, then we should eventually recover when it's working and we attempt to build the crate again.

File: src/utils/daemon.rs
Comment (Nemo157) [inline src/utils/daemon.rs:36]: We should use `{:?}` to print anyhow errors [to include the cause and backtrace if available](https://docs.rs/anyhow/latest/anyhow/struct.Error.html#display-representations) (and can use [captured identifiers](https://blog.rust-lang.org/2022/01/13/Rust-1.58.0.html#captured-identifiers-in-format-strings)).

```suggestion
            log::error!(
                "queue locked because of invalid last_seen_index_reference in database: {err:?}",
            );
```

File: src/utils/mod.rs
Comment (Nemo157) [inline src/utils/mod.rs:53]: Rather than `impl Into<Value>` this could be `impl Serialize` and use [`serde_json::to_value`](https://docs.rs/serde_json/latest/serde_json/fn.to_value.html), that would support nested types if we ever want to store something more than a simple string/bool in the future (like the previous `release_activity` data).

File: src/utils/daemon.rs
Comment (syphar) [reply src/utils/daemon.rs]: good catch! I typically do this (`{:?}`), this time I missed it.

I fixed it.

File: src/utils/mod.rs
Comment (syphar) [reply src/utils/mod.rs:53]: I wasn't sure about the best API for these two methods, but you're right, `impl Serialize` could be the better one, both ways. 

I'll update it later today

**Comment by Nemo157** [comment]
I [threw 3 builders into the docker-compose setup](https://gist.github.com/Nemo157/b8e4a14643eb2a7cffbee4f3c323d0ce) and tested building some 20 crates locally, the queueing handled them all fine.

I wonder if we should record the hostname or something into `builds` too, so we can identify which worker built the docs if we need to investigate issues.

File: src/utils/mod.rs
Comment (syphar) [reply src/utils/mod.rs]: I updated the methods with an idea, I'm happy to take any feedback!

**Comment by syphar** [comment]
@jyn514 @Nemo157 I added new commits for the proposed changes, sot his is ready for another review

**Comment by syphar** [comment]
@Nemo157 @jyn514 ping?

**Review by Nemo157 — APPROVED** [review]
LGTM, and worked well in local testing (I just retested with the hostname changes).

**Comment by syphar** [comment]
I'll merge & deploy this separately after the current merges are live & safe, so I can watch & eventually revert better

**Pull request merged**
